System/Console/Haskeline/Echo.hs

```diff
@@ -0,0 +1,147 @@
+-- computation.
+--
+-- @
+-- bracketInputEcho input action =
+--  'bracket' ('liftIO' $ 'getInputEchoState' input)
+--            ('liftIO' . 'setInputEchoState' input)
+--            (const action)
+-- @
+bracketInputEcho :: MonadException m => Handle -> m a -> m a
+bracketInputEcho input action =
+  bracket (liftIO $ getInputEchoState input)
+          (liftIO . setInputEchoState input)
+          (const action)
+
+-- | Perform a computation with the handle's input echoing disabled. Before
+-- running the computation, the handle's input 'EchoState' is saved, and the
+-- saved 'EchoState' is restored after the computation finishes.
+--
+-- @
+-- withoutInputEcho input action =
+--   'bracketInputEcho' input ('liftIO' ('setInputEchoState' input 'echoOff') >> action)
+-- @
+withoutInputEcho :: MonadException m => Handle -> m a -> m a
+withoutInputEcho input action =
+  bracketInputEcho input (liftIO (setInputEchoState input echoOff) >> action)
+
+-- | Create an @stty@ process, wait for it to complete, and return its output.
+sttyRaw :: Handle -> String -> IO STTYSettings
+sttyRaw input arg = do
+  let stty = (shell $ "stty " ++ arg) {
+        std_in  = UseHandle input
+      , std_out = CreatePipe
+      }
+  (_, mbStdout, _, rStty) <- createProcess stty
+  exStty <- waitForProcess rStty
+  case exStty of
+    e@ExitFailure{} -> throw e
+    ExitSuccess     -> maybe (return "") hGetContents mbStdout
+
+-- | A representation of the terminal input's current echoing state. Example
+-- values include 'echoOff' and 'echoOn'.
+data EchoState
+  = MinTTY STTYSettings
+    -- ^ The argument to (or value returned from) an invocation of the @stty@
+    -- command-line utility. Most POSIX-like shells have @stty@, including
+    -- MinTTY on Windows. Since neither 'hGetEcho' nor 'hSetEcho' work on
+    -- MinTTY, when 'getInputEchoState' runs on MinTTY, it returns a value
+    -- built with this constructor.
+    --
+    -- However, native Windows consoles like @cmd.exe@ or PowerShell do not
+    -- have @stty@, so if you construct an 'EchoState' with this constructor
+    -- manually, take care not to use it with a native Windows console.
+  | DefaultTTY Bool
+    -- ^ A simple on ('True') or off ('False') toggle. This is returned by
+    -- 'hGetEcho' and given as an argument to 'hSetEcho', which work on most
+    -- consoles, with the notable exception of MinTTY on Windows. If you
+    -- construct an 'EchoState' with this constructor manually, take care not
+    -- to use it with MinTTY.
+  deriving (Eq, Ord, Show)
+
+-- | Indicates that the terminal's input echoing is (or should be) off.
+echoOff :: EchoState
+echoOff = if minTTY then MinTTY "-echo" else DefaultTTY False
+
+-- | Settings used to configure the @stty@ command-line utility.
+type STTYSettings = String
+
+-- | Is the current process attached to a MinTTY console (e.g., Cygwin or MSYS)?
+minTTY :: Bool
+#if defined(MINGW)
+minTTY = unsafePerformIO $ do
+  h <- getStdHandle sTD_INPUT_HANDLE
```

Comment:
OK. But keep in mind that Haskell `Handle`s and Win32 `HANDLE`s (the latter of which `isMinTTYHandle` accepts as an argument) are not the same thing, and you have to do a bit of an elaborate dance to convert from the former to the latter. This would require bringing in [some more code](https://github.com/haskell/win32/commit/729f9029e7a1fb8274c0b7f8065dec45f003405a#diff-0546d95d825d023b9c8315fd6043612eR253), which I imagine will be in `Win32-2.6`.

Given your comments [here](https://github.com/judah/haskeline/pull/52#discussion_r98361809), would you be OK with putting this much code into `haskeline`? Or do you want to wait until `Win32-2.6`'s release so that you can import both `isMinTTYHandle` _and_ `withHandleToHANDLE` from `Win32` directly?